### ISSUE TITLE
[2019 ICCV] Attention Augmented Convolutional Networks

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#39]
Convolution operation은 limited receptive field의 locality와 weight sharing을 통한 translation equivariance라는 특성을 가짐

해당 2가지 특성으로 인해 CNN은 crucial inductive bias를 가지게 되어 image의 global context를 capture하는 것이 어려움

이를 convolution에 self-attention을 추가하여 성능 개선을 한 논문

어떻게 작동하는지와 결과 figures만 설명

## 1. Attention Augmentation
![image](https://user-images.githubusercontent.com/49643709/137822807-aec6fc2d-6ff2-4447-9ed8-525f2175a495.png)

### 1.1. Self-attention over images
<img src='https://user-images.githubusercontent.com/49643709/137823735-0e3c0a72-d370-4b87-bdb3-f0a814acae08.png' width=70% height=70%>

1. (N, C, H, W)의 input tensor를 (N, C, H * W) 형태로 flatten
2. Flatten input을 X라고 하면 query matrix, key matrix, value matrix를 통해 여러 head들에 대하여 output을 구함
3. Head별로 나온 output들을 concat한 후, linear transformation weight matrix를 통해 shape를 맞춰줌

### 1.2. Two-dimensional relative positional embeddings
<img src='https://user-images.githubusercontent.com/49643709/137824757-78347d85-0c36-4e84-b799-8b5f3e83520c.png' width=70% height=70%>

Position에 대한 정보가 없다면, self-attention은 permutation equivariant한데, 이는 이미지와 같은 데이터에 적합하지 않음

기존의 image에서 사용되었던 positional encoding 방법들은 permutation equivariant를 방지할 수 있으나 translation equivariance하다는 특성을 가지지 못함

Permutation equivariant를 방지하며 translation equivariance 특성을 가져오기 위해, 저자들은 two dimension에서 relative position encodings를 통해 해결함

1. Pixel i, j의 row와 column의 차이를 각각 계산 (2*H-1, 2*W-1인 이유)
2. Row와 column weight matrix를 통해 row와 column의 relative position logits를 생성 (relative positional embeddings)

참고로 row와 column relative positional embeddings는 heads 내에서 share됨
(relative positional embeddings are shared across heads but not layers)

### 1.3. Attention augmented convolution
<img src='https://user-images.githubusercontent.com/49643709/137828344-0a01801d-23c9-4ca8-97cd-7a7d354ca6cb.png' width=70% height=70%>

1. Input이 주어졌을 때, channel 수를 2개로 나눔 (나누는 비율은 hyperparameter)
2. 나눈 파트 중 1개는 convolution operation을 사용하고, 나머지 1개는 self-attention을 사용
3. 각 파트에서 나온 output을 concat

### 1.4. Attention augmented convolutional architecture

- Attention augmented convolution 뒤에 batchnorm layer 사용
- Residual block 하나당 attention augmented convolution 사용
- 메모리 문제로 인해, last layer에서 attention augmented convolution 적용하며 small batch size 사용
- Input spatial dimension이 큰 경우 메모리 문제로 인해 다음과 같이 적용
    1. 3*3 kernel with stride 2 average pooling을 통해 downsample
    2. Downscaled input에 self-attention 사용
    3. Bilinear interpolation을 통해 upsampling

### 2.1. Image Classification & Detection
<img src='https://user-images.githubusercontent.com/49643709/138017080-cffbf68c-136a-47a9-aa86-6b5c83ccca14.png' width=70% height=70%>
<img src='https://user-images.githubusercontent.com/49643709/138017350-181f47e9-49ea-4c26-8781-321ba03718e0.png' width=70% height=70%>
<img src='https://user-images.githubusercontent.com/49643709/138017442-acf65350-640f-4852-94e7-f30c4c833290.png' width=70% height=70%>
<img src='https://user-images.githubusercontent.com/49643709/138017559-3cb94940-6414-451a-bc34-e5227a26d00c.png' width=70% height=70%>
<img src='https://user-images.githubusercontent.com/49643709/138017545-b85ce366-1701-4975-a2ed-bac947e8351e.png' width=70% height=70%>
<img src='https://user-images.githubusercontent.com/49643709/138017688-a39d2f19-c555-46a7-8c4b-f2e2477f0463.png'>

#### 2.2.1. Importance of Attention Augmentation
<img src='https://user-images.githubusercontent.com/49643709/138017776-cd339f4d-cf87-427f-9502-f19c9f09a95a.png' width=70% height=70%>
<img src='https://user-images.githubusercontent.com/49643709/138017795-1cf34f76-4235-4d41-bdc8-2a36c06b58e1.png' width=70% height=70%>

- k : ratio of key depth to number of original output filters
- v : ratio of attentional channels to number of original output filters

#### 2.2.2. Importance of position encodings
<img src='https://user-images.githubusercontent.com/49643709/138018892-99838a7c-7779-4fcd-8a30-9ab0844f8e2d.png' width=70% height=70%>
<img src='https://user-images.githubusercontent.com/49643709/138018912-62124733-8ea6-49b8-8324-7f1dee133253.png' width=70% height=70%>
